crates/integrations/datafusion/src/catalog.rs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{any::Any, sync::Arc};
+
+use dashmap::DashMap;
+use datafusion::catalog::{schema::SchemaProvider, CatalogProvider};
+use futures::future::try_join_all;
+use iceberg::{Catalog, NamespaceIdent, Result};
+
+use crate::schema::IcebergSchemaProvider;
+
+/// Provides an interface to manage and access multiple schemas
+/// within an Iceberg [`Catalog`].
+///
+/// Acts as a centralized catalog provider that aggregates
+/// multiple [`SchemaProvider`], each associated with distinct namespaces.
+pub struct IcebergCatalogProvider {
+    /// A concurrent `HashMap` where keys are namespace names
+    /// and values are dynamic references to objects implementing the
+    /// [`SchemaProvider`] trait.
+    schemas: DashMap<String, Arc<dyn SchemaProvider>>,
```

Comment:
Would a std Hashmap suffice? I don't see the concurrency properties of DashMap being used.

I might be missing something though.

Reply:
the datafusion docs provide a link to a reference implementation by [delta-rs](https://github.com/delta-io/delta-rs/blob/951436ecec476ce65b5ed3b58b50fb0846ca7b91/crates/deltalake-core/src/data_catalog/unity/datafusion.rs#L21), that uses a DashMap as well. 

My understanding based on the [docs](https://docs.rs/datafusion/latest/datafusion/catalog/trait.CatalogProvider.html#implementing-remote-catalogs) is that datafusion might share the `CatalogProvider` on multiple threads inside an Arc; thus concurrently accessing our "cached" schemas, etc.? Perhaps someone else can clarify this?

> [...] The pattern that DataFusion itself uses to plan SQL queries is to walk over the query to [find all schema / table references in an async function](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionState.html#method.resolve_table_references), performing required remote catalog in parallel, and then plans the query using that snapshot.

Reply:
Will the map ever be mutated?
It might be, I was just not seeing it in this PR.

If not, then a std Hashmap seems sufficient to me.

Reply:
Cool, well then I think a std Hashmap should be fine, since no actual concurrent operations need to happen